lib/cloak/fields/pbkdf2.ex

```diff
@@ -0,0 +1,193 @@
+  always results in the same value, while secure encryption does not. Be
+  warned, however, that hashing will expose which fields have the same value,
+  because they will contain the same hash.
+
+  ## Configuration
+
+  Create a `PBKDF2` field in your project:
+
+      defmodule MyApp.Hashed.PBKDF2 do
+        use Cloak.Fields.PBKDF2, otp_app: :my_app
+      end
+
+  Then, configure it with a `:secret`, an `:algorithm`, the maximum `:size`
+  of the stored key (in bytes), and a number of `:iterations`, either using
+  mix configuration:
+
+      config :my_app, MyApp.Hashed.PBKDF2,
+        algorithm: :sha256,
+        iterations: 10_000,
+        secret: "secret"
+
+  Or using the `init/1` callback to fetch configuration at runtime:
+
+      defmodule MyApp.Hashed.PBKDF2 do
+        use Cloak.Fields.PBKDF2, otp_app: :my_app
+
+        @impl Cloak.Fields.PBKDF2
+        def init(config) do
+          config = Keyword.merge(config, [
+            algorithm: :sha256,
+            iterations: 10_000,
+            secret: System.get_env("PBKDF2_SECRET")
+          ])
+
+          {:ok, config}
+        end
+      end
+
+  ## Usage
+
+  Create the hash field with the type `:binary`. Add it to your schema
+  definition like this:
+
+      schema "table" do
+        field :field_name, MyApp.Encrypted.Binary
+        field :field_name_hash, MyApp.Hashed.PBKDF2
+      end
+
+  Ensure that the hash is updated whenever the target field changes with the
+  `put_change/3` function:
+
+      def changeset(struct, attrs \\\\ %{}) do
+        struct
+        |> cast(attrs, [:field_name, :field_name_hash])
+        |> put_hashed_fields()
+      end
+
+      defp put_hashed_fields(changeset) do
+        changeset
+        |> put_change(:field_name_hash, get_field(changeset, :field_name))
+      end
+
+  Query the Repo using the `:field_name_hash` in any place you would typically
+  query by `:field_name`.
+
+      user = Repo.get_by(User, email_hash: "user@email.com")
+  """
+
+  @typedoc "Digest algorithms supported by Cloak.Field.PBKDF2"
+  @type algorithms :: :md4 | :md5 | :ripemd160 | :sha | :sha224 | :sha256 | :sha384 | :sha512
+
+  @doc """
+  Configures the `PBKDF2` field using runtime information.
+
+  ## Example
+
+      @impl Cloak.Fields.PBKDF2
+      def init(config) do
+        config = Keyword.merge(config, [
+          algorithm: :sha256,
+          secret: System.get_env("PBKDF2_SECRET")
+        ])
+
+        {:ok, config}
+      end
+  """
+  @callback init(config :: Keyword.t()) :: {:ok, Keyword.t()} | {:error, any}
+
+  @doc false
+  defmacro __using__(opts) do
+    otp_app = Keyword.fetch!(opts, :otp_app)
+
+    quote do
+      @behaviour Cloak.Fields.PBKDF2
+      @behaviour Ecto.Type
+      @algorithms ~w[
+        md4
+        md5
+        ripemd160
+        sha
+        sha224
+        sha256
+        sha384
+        sha512
+      ]a
+      @sizes [16, 32, 64, 128]
```

Comment:
Limiting the size of the blind index to specific byte lengths is an optimization for space within the database. I like the idea of same-size index keys. What do you think about this? There is also a version of the function that doesn't have a specified derived key length, so we can remove this easily.

Reply:
I think we should support the `:size` option, but I don't see a reason to limit the options. Let people choose whatever length they prefer.